Fix ParameterExpression.gradient() to return numeric values for constant derivatives

Aqil-Ahmad opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.
- [ ] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

## Summary
Fixes #14845:
where `ParameterExpression.gradient()` returns a `ParameterExpression` object instead of the expected numeric value for constant derivatives.

## Details and comments

### Problem
After PR #14757, the behavior changed and `x.gradient(x)` started returning a `ParameterExpression` object instead of the expected `1.0` value.

**Before:**
```python
>>> x.gradient(x)
<qiskit._accelerate.circuit.ParameterExpression object at 0x...>
```

### Solution
Modified `gradient()` method to:
- Check if derivative is constant using `d_expr.expr.parameters().is_empty()`
- Return native Python types (float, int, complex) for constant derivatives
- Return `ParameterExpression` for symbolic derivatives

**After:**
```python
>>> x.gradient(x)
1.0
```

Aqil-Ahmad committed:
Fix: ParameterExpression.gradient() to return numeric values for constants

garrison commented:
Thank you for submitting a fix. Would you be able to add a test as well?

Aqil-Ahmad commented:
> Thank you for submitting a fix. Would you be able to add a test as well?

Sure! I’d be happy to add a test.

Aqil-Ahmad committed:
Add test and fix some issues(lint etc)

garrison commented on test/python/circuit/test_parameter_expression.py:
```diff
@@ -15,16 +15,13 @@
-
-from test import combine
-from test import QiskitTestCase
+from test import QiskitTestCase, combine
```
I suspect it may be quicker/easier to review this PR if the only changes to this file are addition of tests and any new imports that might be necessary. The unrelated changes should be kept to a minimum to enable speedy review.

Aqil-Ahmad replied:
Got it. I’ll remove the unrelated changes.

Cryoris commented on test/python/circuit/test_parameter_expression.py:
```diff
@@ -467,6 +467,146 @@ def test_sign_derivative_errors(self):
         with self.assertRaises(RuntimeError):
             _ = expr.gradient(x)
 
+    def test_gradient_constant_derivatives(self):
```
A few of these tests seem unrelated to the bugfix, could you remove them and only keep ones that check the fixed behavior?

Aqil-Ahmad replied:
I have removed all other tests and kept only the cases where the derivative is numeric.

Cryoris reviewed (dismissed):
LGTM, thank you @Aqil-Ahmad! No reno needed as the bug is not yet released.

garrison commented on test/python/circuit/test_parameter_expression.py:
```diff
@@ -467,6 +467,27 @@ def test_sign_derivative_errors(self):
+        x = Parameter("x")
+        y = Parameter("y")
+
+        test_cases = [
+            (x, x, 1.0),
+            (x + 0, x, 1.0),
+            (0 * x, x, 0.0),
+            (x / 2, x, 0.5),
+            (x - x, x, 0.0),
+            (5 + x - x, x, 0.0),
+            (2 * x + y - x, x, 1.0),
+        ]
```
This is great, thank you.  ~~At this point, the main way I think this PR could be improved is to move these outside the function and use the `@ddt.data` and `ddt.unpack` decorators, similar to at the following link:~~

https://github.com/Qiskit/qiskit/blob/14df5941a5915a2b517cdf9bc014a6838e752f92/test/python/synthesis/xx_decompose/test_polytopes.py#L38-L45

garrison replied:
Nevermind, I realize now that it's not straightforward to do this because `x` and `y` only exist in the function body.  I don't know enough about ddt to know if a fixture is a better way to do it.  The current code may be the best way to do this.

garrison commented on test/python/circuit/test_parameter_expression.py:
```diff
@@ -467,6 +467,29 @@ def test_sign_derivative_errors(self):
+            with self.subTest(expr=str(expr), param=str(param)):
+                result = expr.gradient(param)
+                if isinstance(result, str):
+                    result = float(result)
```
Why might we expect the gradient to return a string rather than a numeric value or `ParameterExpression`?

Aqil-Ahmad replied:
Some CI checks were failing with the error` '1' is not an instance of (float, int, complex)`. Locally, all the test cases return floats, so I wasn’t able to reproduce it. I added a condition to normalize the result to a float, but the checks are still failing with the same mismatch error. This might be due to a version/environment difference, and I’d appreciate some guidance on the best way to handle it.

Cryoris committed:
Fix conversion to numeric value

The argument in try_to_value must be set to true for this PR to work. If false then the code will never exercise the path for built-in numerics.

Also simplify the results, no additional Ok(..?) is needed.
